pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/ReaderThread.java

```diff
@@ -99,11 +100,17 @@ public void run() {
 
             log.info("Starting to read {} with reader thread {}", topicRange, getName());
 
+            Watermark lastWatermark = null;
             while (running) {
                 Message<T> message = reader.readNext(pollTimeoutMs, TimeUnit.MILLISECONDS);
                 if (message != null) {
                     emitRecord(message);
                 }
+                Watermark watermark = reader.getLastWatermark();
+                if (watermark != null && !watermark.equals(lastWatermark)) {
+                    emitWatermark(watermark.getEventTime());
+                    lastWatermark = watermark;
+                }
```

Comment:
The mechanics of getting watermarks from the Pulsar reader is, after `read` completes, check for an updated watermark, and emit as necessary.  The `read` method must have a shorter timeout than usual, otherwise watermarks would be delayed until the read completes.

Note that `readAsync` has new semantics; any outstanding read is completed immediately (with `null`) when a watermark arrives.  But the Flink code doesn't use `readAsync` so we went with this approach.

More info on the consumer/reader API changes:
https://github.com/EronWright/pulsar-hackathon-2021-projects-watermarking/blob/main/README.md#how-it-works